includes/classes/Indexable/Post/Post.php

```diff
@@ -404,118 +420,120 @@ private function prepare_terms( $post ) {
 			return [];
 		}
 
-		$terms = [];
+		$object_terms = wp_get_object_terms( $post->ID, wp_list_pluck( $selected_taxonomies, 'name' ), [ 'update_term_meta_cache' => false ] );
+		if ( empty( $object_terms ) || is_wp_error( $object_terms ) ) {
+			return [];
+		}
 
 		/**
-		 * Filter to allow child terms to be indexed
+		 * Filter to allow parent terms to be indexed.
 		 *
 		 * @hook ep_sync_terms_allow_hierarchy
-		 * @param  {bool} $allow True means allow
+		 * @param {bool} $allow True means allow
 		 * @return  {bool} New value
 		 */
 		$allow_hierarchy = apply_filters( 'ep_sync_terms_allow_hierarchy', true );
 
-		foreach ( $selected_taxonomies as $taxonomy ) {
-			// If we get a taxonomy name, we need to convert it to taxonomy object
-			if ( ! is_object( $taxonomy ) && taxonomy_exists( (string) $taxonomy ) ) {
-				$taxonomy = get_taxonomy( $taxonomy );
+		// Build out a flat array with all the term data we will be indexing.
+		$terms_dictionary = [];
+		foreach ( $object_terms as $term ) {
+			if ( ! isset( $terms_dictionary[ $term->term_taxonomy_id ] ) ) {
+				$terms_dictionary[ $term->term_taxonomy_id ] = [
+					'term_id'          => $term->term_id,
+					'slug'             => $term->slug,
+					'name'             => $term->name,
+					'parent'           => $term->parent,
+					'term_taxonomy_id' => $term->term_taxonomy_id,
+					'taxonomy'         => $term->taxonomy, // Note: This is removed before it's indexed.
```

Comment:
Let's adjust the note to include the reason for it being added in the first place.

Reply:
Hmm, actually seeing a way to avoid this and simplify a bit.

Reply:
Avoided 👍